### PR TITLE
[ PECO - 1768 ] PySQL: adjust HTTP retry logic to align with Go and Nodejs drivers

### PR DESCRIPTION
**Pull request by jprakash-db**
## Description

Adjusted the retry logic to be in align with the Go and Node JS drivers.

- Implemented the exponential backoff algorithm for backoff time calculation
- Updated the number of retries and and the default value ranges for the delay between retries

## Tests

Added unit tests for testing different functionalities

### CODE REVIEW

**Comment by jackyhu-db** [comment]
Please include the unit test.

File: src/databricks/sql/thrift_backend.py [hunk 1]
```diff
@@ -64,8 +64,8 @@
 # - 900s attempts-duration lines up w ODBC/JDBC drivers (for cluster startup > 10 mins)
 _retry_policy = {  # (type, default, min, max)
     "_retry_delay_min": (float, 1, 0.1, 60),
-    "_retry_delay_max": (float, 60, 5, 3600),
-    "_retry_stop_after_attempts_count": (int, 30, 1, 60),
+    "_retry_delay_max": (float, 30, 5, 3600),
```
Comment (jackyhu-db) [inline src/databricks/sql/thrift_backend.py:67]: why do you change from 60 to 30?

**Comment by jackyhu-db** [comment]
Please update PR description to be more specific (e.g. change the retry logic to use exponential backoff algorithm)

**Review by gopalldb — COMMENTED** [review]
Looks good, add tests

**Review by jackyhu-db — APPROVED** [review]

**Review by gopalldb — APPROVED** [review]

**Pull request merged**